metadata/metadata.go

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (C) 2017 Red Hat, Inc.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package metadata
+
+import (
+	"reflect"
+	"sync"
+
+	"github.com/skydive-project/skydive/api"
+	"github.com/skydive-project/skydive/logging"
+	"github.com/skydive-project/skydive/topology"
+	"github.com/skydive-project/skydive/topology/graph"
+)
+
+type UserMetadataManager struct {
+	sync.RWMutex
+	graph.DefaultGraphListener
+	graph           *graph.Graph
+	metadataHandler *api.UserMetadataAPIHandler
+	metadatas       map[string]*api.UserMetadata
+	watcher         api.StoppableWatcher
+}
+
+func (u *UserMetadataManager) OnNodeAdded(n *graph.Node) {
+	u.nodeEvent()
+}
+
+func (u *UserMetadataManager) OnNodeUpdated(n *graph.Node) {
+	u.nodeEvent()
+}
+
+func (u *UserMetadataManager) nodeEvent() {
+	for _, m := range u.metadatas {
+		u.addUserMetadata(m)
+	}
+}
+
+func (u *UserMetadataManager) applyGremlinExpr(query string) []interface{} {
+	res, err := topology.ExecuteGremlinQuery(u.graph, query)
+	if err != nil {
+		logging.GetLogger().Errorf("Gremlin error: %s", err.Error())
+		return nil
+	}
+	return res.Values()
+}
+
+func (u *UserMetadataManager) addUserMetadata(umd *api.UserMetadata) {
+	u.Lock()
+	u.metadatas[umd.UUID] = umd
+	u.Unlock()
+
+	nodes := u.applyGremlinExpr(umd.GremlinQuery)
+	for _, node := range nodes {
+		var n *graph.Node
+		switch node.(type) {
+		case *graph.Node:
+			n = node.(*graph.Node)
+		default:
+			continue
+		}
+		metadata := n.Metadata()
+		if usermd, ok := metadata["UserMetadata"]; !ok {
+			userdata := make(graph.Metadata)
+			userdata[umd.Key] = umd.Value
+			u.graph.AddMetadata(node, "UserMetadata", userdata)
```

Comment:
I would prefer to not have nested metadata, just use AddMetadata with UserMetadata as prefix for the key, moreover it will allow you to not check for the key